File: scripts/lib/flows.py
"""
Flow construction: turn the resolved dependency graph into per-entry-point
"flows" (forward call/dependency chains), with high-in-degree "commons" nodes
treated as non-expandable leaves so a flow never explodes into the whole graph.

Language-agnostic: operates only on Declaration objects (for hub marking) and
plain edge dicts {from, to, kind} (for traversal). Mirrors the client-side
trace in viewer/index.html — keep the pruning rules in sync.
"""
from __future__ import annotations
from collections import defaultdict, deque
from .extractors.base import Declaration


def mark_hubs(declarations: list[Declaration], percentile: float = 0.05) -> set[str]:
    """Mark the top `percentile` of declarations by in-degree as hubs.

    Hubs are the "everyone calls this" nodes (logger/util/common base class);
    flow traversal shows them but does not expand them, so a flow stays about
    what is *specific* to its path. Rank-based and gated on in_degree > 0 so a
    small repo (where the percentile boundary lands in the zero block) gets no
    hubs at all. Reads the `_in_degree` cached by core.build_graph. Sets `_hub`
    on every declaration and returns the set of hub qualified_names.
    """
    for d in declarations:
        d._hub = False  # type: ignore[attr-defined]
    nonzero = [d for d in declarations if getattr(d, "_in_degree", 0) > 0]
    # Sort by in-degree desc, then qualified_name asc for deterministic ties.
    nonzero.sort(key=lambda d: (-d._in_degree, d.qualified_name))  # type: ignore[attr-defined]
    k = int(len(nonzero) * percentile)
    hub_ids: set[str] = set()
    for d in nonzero[:k]:
        d._hub = True  # type: ignore[attr-defined]
        hub_ids.add(d.qualified_name)
    return hub_ids


def trace_flow(seed: str, adjacency: dict[str, list[str]], hub_ids: set[str],
               max_depth: int = 6) -> tuple[list[str], list[dict]]:
    """Forward BFS from `seed` over `adjacency` (from → [to], built from
    'uses' edges). A hub node is included but not expanded (leaf) — unless it
    is the seed itself. Each node is placed once, at its shortest depth; edges
    back to an already-placed node are omitted (keeps the pipeline a readable
    tree). Stops past `max_depth`. Returns (ordered node ids, edge dicts).
    """
    visited = {seed}
    order = [seed]
    flow_edges: list[dict] = []
    depth = {seed: 0}
    q = deque([seed])
    while q:
        u = q.popleft()
        if u != seed and u in hub_ids:
            continue                      # hub: leaf, do not expand
        if depth[u] >= max_depth:
            continue
        for v in adjacency.get(u, []):
            if v in visited:
                continue                  # already placed — omit cross/back edge
            visited.add(v)
            depth[v] = depth[u] + 1
            order.append(v)
            flow_edges.append({"from": u, "to": v})
            q.append(v)
    return order, flow_edges


def build_flows(seeds: list[str], declarations: list[Declaration],
                edges: list[dict], hub_ids: set[str],
                max_depth: int = 6) -> list[dict]:
    """Build one deterministic candidate flow per seed (an entry-point
    qualified_name). Traverses 'uses' edges only. Skips seeds that are not
    real declarations. An entry point with no outgoing edges yields a valid
    single-node flow (an honest "no downstream calls resolved" result).
    """
    adjacency: dict[str, list[str]] = defaultdict(list)
    for e in edges:
        if e.get("kind") == "uses":
            adjacency[e["from"]].append(e["to"])

    by_qname = {d.qualified_name: d for d in declarations}
    out: list[dict] = []
    seen_seeds: set[str] = set()
    for seed in seeds:
        if seed in seen_seeds:
            continue                      # dedup: one flow per id, even if a name collides
        seen_seeds.add(seed)
        decl = by_qname.get(seed)
        if decl is None:
            continue
        nodes, fedges = trace_flow(seed, adjacency, hub_ids, max_depth)
        out.append({
            "id": "flow:" + seed,
            "name": decl.name,
            "description": "",
            "seed": seed,
            "nodes": nodes,
            "edges": fedges,
            "confidence": "high",
        })
    return out

File: scripts/lib/flows.mjs
// scripts/lib/flows.mjs — flow construction (forward call/dependency chains per
// entry point, with high-in-degree hubs as non-expandable leaves). Port of flows.py.
import { qualifiedName } from './extractors/base.mjs';

/** 把 supertype / ref 字符串归一化为短名，镜像 core.resolve 的规则：
 *  先剥泛型 <...> 与调用括号 (...)，再取 . :: / 的末段。 */
export function shortName(raw) {
  const base = String(raw || '').split('<')[0].split('(')[0];
  return base.split('.').pop().split('::').pop().split('/').pop();
}

/** Map<短名, 实现声明[]>：仅保留 >=2 个实现的接口/抽象类型，桶内按
 *  (_importance desc, qualified_name asc) 排序。从声明的 supertypes 字符串
 *  构建，而非图的 extends 边——后者在目标节点未被抽取时已被 core.resolve
 *  丢弃（okhttp 的 fun interface Interceptor 即如此）。语言无关。 */
export function buildDispatchIndex(declarations) {
  const byShort = new Map();
  for (const d of declarations) {
    for (const s of d.supertypes || []) {
      const k = shortName(s);
      if (!k) continue;
      if (!byShort.has(k)) byShort.set(k, []);
      byShort.get(k).push(d);
    }
  }
  const index = new Map();
  for (const [k, impls] of byShort) {
    if (impls.length < 2) continue;
    impls.sort((a, b) => {
      const ia = a._importance || 0, ib = b._importance || 0;
      if (ib !== ia) return ib - ia;
      const qa = qualifiedName(a), qb = qualifiedName(b);
      return qa < qb ? -1 : qa > qb ? 1 : 0;
    });
    index.set(k, impls);
  }
  return index;
}

/** Mark the top `percentile` of declarations by in-degree as hubs. Sets _hub on
 *  every declaration; returns the set of hub qualified_names. */
export function markHubs(declarations, percentile = 0.05) {
  for (const d of declarations) d._hub = false;
  const nonzero = declarations.filter((d) => (d._in_degree || 0) > 0);
  nonzero.sort((a, b) => {
    if (b._in_degree !== a._in_degree) return b._in_degree - a._in_degree;
    const qa = qualifiedName(a), qb = qualifiedName(b);
    return qa < qb ? -1 : qa > qb ? 1 : 0;
  });
  const k = Math.floor(nonzero.length * percentile);
  const hubIds = new Set();
  for (let i = 0; i < k; i++) {
    nonzero[i]._hub = true;
    hubIds.add(qualifiedName(nonzero[i]));
  }
  return hubIds;
}

/** Forward BFS from `seed` over uses-adjacency, plus dispatch expansion.
 *  At a node U, after following resolved uses-edges, scan U's raw ref strings
 *  (ctx.refsByQname); any ref whose short-name is a dispatch-index key (an
 *  interface/abstract type with >=2 implementors) fans out to its implementors
 *  as `dispatch` edges, capped at ctx.maxFanout (overflow recorded in omitted).
 *  A hub is a leaf (unless it is the seed). visited-set prevents revisits/cycles.
 *  Returns [orderedIds, edges, omitted]. Pass ctx=null to disable dispatch. */
export function traceFlow(seed, adjacency, hubIds, maxDepth = 6, ctx = null) {
  const visited = new Set([seed]);
  const order = [seed];
  const flowEdges = [];
  const omitted = [];
  const depth = new Map([[seed, 0]]);
  const q = [seed];
  let head = 0;
  while (head < q.length) {
    const u = q[head++];
    if (u !== seed && hubIds.has(u)) continue; // hub: leaf
    if (depth.get(u) >= maxDepth) continue;
    // 1. resolved uses-edges
    for (const v of adjacency.get(u) || []) {
      if (visited.has(v)) continue;
      visited.add(v);
      depth.set(v, depth.get(u) + 1);
      order.push(v);
      flowEdges.push({ from: u, to: v, kind: 'uses' });
      q.push(v);
    }
    // 2. dispatch expansion via unresolved interface refs
    if (ctx && ctx.dispatchIndex) {
      const seenVia = new Set();
      for (const raw of ctx.refsByQname.get(u) || []) {
        const k = shortName(raw);
        if (!k || seenVia.has(k)) continue;
        const impls = ctx.dispatchIndex.get(k);
        if (!impls) continue;
        seenVia.add(k);
        const fresh = impls.filter((d) => !visited.has(qualifiedName(d)));
        const cap = ctx.maxFanout || 8;
        const take = fresh.slice(0, cap);
        for (const d of take) {
          const v = qualifiedName(d);
          visited.add(v);
          depth.set(v, depth.get(u) + 1);
          order.push(v);
          flowEdges.push({ from: u, to: v, kind: 'dispatch', via: k });
          q.push(v);
        }
        if (fresh.length > take.length) {
          omitted.push({ from: u, via: k, count: fresh.length - take.length });
        }
      }
    }
  }
  return [order, flowEdges, omitted];
}

/** Candidate flows: one per seed, traced with dispatch expansion when opts
 *  carries a dispatchIndex. opts = { dispatchIndex, maxFanout, seedKind }.
 *  seedKind: Map<seed, 'entry-point'|'public-orchestrator'>; a public-orchestrator
 *  seed is marked confidence:'candidate' (Phase 2 curates), entry-point stays 'high'. */
export function buildFlows(seeds, declarations, edges, hubIds, maxDepth = 6, opts = {}) {
  const adjacency = new Map();
  for (const e of edges) {
    if (e.kind === 'uses') {
      if (!adjacency.has(e.from)) adjacency.set(e.from, []);
      adjacency.get(e.from).push(e.to);
    }
  }
  const byQname = new Map(declarations.map((d) => [qualifiedName(d), d]));
  const dispatchIndex = opts.dispatchIndex || null;
  const ctx = dispatchIndex
    ? {
        dispatchIndex,
        maxFanout: opts.maxFanout || 8,
        refsByQname: new Map(declarations.map((d) => [qualifiedName(d), d.refs || []])),
      }
    : null;
  const seedKind = opts.seedKind || new Map();
  const out = [];
  const seenSeeds = new Set();
  for (const seed of seeds) {
    if (seenSeeds.has(seed)) continue;
    seenSeeds.add(seed);
    const decl = byQname.get(seed);
    if (decl == null) continue;
    const [nodes, fedges, omitted] = traceFlow(seed, adjacency, hubIds, maxDepth, ctx);
    const kind = seedKind.get(seed) || 'entry-point';
    const flow = {
      id: 'flow:' + seed,
      name: decl.name,
      description: '',
      seed,
      seed_kind: kind,
      nodes,
      edges: fedges,
      confidence: kind === 'public-orchestrator' ? 'candidate' : 'high',
    };
    if (omitted.length) flow.dispatch_omitted = omitted;
    out.push(flow);
  }
  return out;
}
